### PR TITLE
refactor dialing to not panic, and to be smart about ordering

### PR DESCRIPTION
**Pull request by whyrusleeping**
After [yet another goprocess bug](https://ipfs.io/ipfs/QmbYDMxoHfrehLdy5Ai3XZRBacjt3JojVuUUEdpRAHa9mn) I decided to just rip it out. Its gone now. bye bye.

I also figured that, while I was at it, I would add a bit of priority to the order in which we dial addresses. 
First: addresses that dont consume a file descriptor.
Second: localhost addresses, as they should either succeed or fail immediately.
Third: all normal tcp addresses.
Last: dockers '172.17' range of addresses, these are almost always trouble.

License: MIT
Signed-off-by: Jeromy <jeromyj@gmail.com>

### CODE REVIEW

File: p2p/net/swarm/swarm_dial.go [hunk 1]
```diff
@@ -371,87 +373,84 @@ func (s *Swarm) dialAddrs(ctx context.Context, d *conn.Dialer, p peer.ID, remote
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel() // cancel work when we exit func
 
-	foundConn := make(chan struct{})
-	conns := make(chan conn.Conn, len(remoteAddrs))
+	conns := make(chan conn.Conn)
```
Comment (whyrusleeping) [inline p2p/net/swarm/swarm_dial.go]: if this gets buffered, then open connections just sit in the channel and never get closed

File: p2p/net/swarm/addr/addr.go [hunk 2]
```diff
@@ -188,10 +190,25 @@ func ResolveUnspecifiedAddresses(unspecAddrs, ifaceAddrs []ma.Multiaddr) ([]ma.M
 	return outputAddrs, nil
 }
 
+const interfaceAddrCacheLife = time.Second * 5
+
+var interfaceAddrLastCache time.Time
+var interfaceAddrCache []ma.Multiaddr
+var interfaceAddrLock sync.Mutex
```
Comment (jbenet) [inline p2p/net/swarm/addr/addr.go]: :-1: to static caches like this. make it an object instead:

```go
netAddrCache := NewInterfaceAddrCache()
netAddrCache.InterfaceAddresses()
```
and have a "DefaultCache" -- the way `http` lib does. compartmentalizes and cleans all this stuff.

File: p2p/net/swarm/swarm_dial.go [hunk 3]
```diff
@@ -533,3 +532,63 @@ func isTCPMultiaddr(a ma.Multiaddr) bool {
...
+		if isFDCostlyTransport(b) {
+			return true
+		}
+
+		// if neither consume fd's, assume equal ordering
+		return false
+	}
+
+	// dial localhost addresses next, they should fail immediately
+	if manet.IsIPLoopback(a) {
```
Comment (jbenet) [inline p2p/net/swarm/swarm_dial.go]: localhost utp addresses should be first, not after internet utp addresses.

File: p2p/net/swarm/swarm_dial.go [hunk 4]
```diff
@@ -533,3 +532,63 @@ func isTCPMultiaddr(a ma.Multiaddr) bool {
...
+	}
+
+	if parts[1] == "ip4" && strings.HasPrefix(parts[2], "172.17.") {
+		return true
+	}
+
+	return false
+}
+
+type AddrList []ma.Multiaddr
```
Comment (jbenet) [inline p2p/net/swarm/swarm_dial.go]: it's a good idea but this should really be the work of an estimator, and a smart dialer thing that can maybe learn which types of addresses wont work.

Comment (jbenet) [inline p2p/net/swarm/swarm_dial.go]: also, this belongs in a separate PR.

File: p2p/net/swarm/swarm_dial.go [hunk 5]
```diff
@@ -371,87 +373,84 @@ func (s *Swarm) dialAddrs(ctx context.Context, d *conn.Dialer, p peer.ID, remote
...
-				})
+		limiter := make(chan struct{}, 8)
+		// permute addrs so we try different sets first each time.
+		for _, addr := range remoteAddrs {
+			select {
+			case <-ctx.Done(): // our context was cancelled
+				return
+			case limiter <- struct{}{}:
+				// take the token, move on
+			}
```
Comment (jbenet) [inline p2p/net/swarm/swarm_dial.go]: i believe the `limiter` should be grabbed below, not here

File: p2p/net/swarm/swarm_dial.go [hunk 6]
```diff
@@ -371,87 +373,84 @@ func (s *Swarm) dialAddrs(ctx context.Context, d *conn.Dialer, p peer.ID, remote
...
+				// take the token, move on
 			}
-		})
 
-		processctx.CloseAfterContext(limiter, ctx)
+			// we have to do the waiting concurrently because there are addrs
+			// that SHOULD NOT be rate limited (utp), nor blocked by other
+			// rate limited addrs (tcp).
+			go func(rlc <-chan struct{}, a ma.Multiaddr) {
+				dialSingleAddr(a)
```
Comment (jbenet) [inline p2p/net/swarm/swarm_dial.go]: i believe `limiter` should be grabbed right here, right before `dialSingleAddr(a)`

**Comment by jbenet** [comment]
weird, not sure why the goproc limiter has a bug. also this is not a goprocess bug it's a goprocess/limiter bug-- it's UX is bad.

(we can change the goprocess interface to not panic on this sort of thing, and instead close the process immediately.)

File: p2p/net/swarm/swarm_dial.go (hunk 3 above)
Comment (whyrusleeping) [inline p2p/net/swarm/swarm_dial.go]: ah, good point.

File: p2p/net/swarm/swarm_dial.go (hunk 5 above)
Comment (whyrusleeping) [inline p2p/net/swarm/swarm_dial.go]: so we should take the fd limiter before we take the per-peer dial limiter? I guess that does make sense.

**Comment by whyrusleeping** [comment]
alright, stripped this PR down to just the panic fix simplification stuff.

**Comment by whyrusleeping** [comment]
huh, maybe I should fix this one too: https://travis-ci.org/ipfs/go-ipfs/jobs/85398391

Since it decided to fail on this PR of any it could have chosen

**Comment by jbenet** [comment]
@whyrusleeping ok sounds good.
